gryphon/lib/exchange/binance_je.py

```diff
@@ -0,0 +1,188 @@
+
+logger = get_logger(__name__)
+
+
+def binance_signature(params, secret):
+    """Signature of a url query string as required by Binance
+
+    Parameters
+    ----------
+    params : dict
+        query string parameters from a url as given by a request object
+    secret : str
+        binance user secret for signing the query string
+
+    Returns
+    -------
+    str
+    """
+    query_string = ""
+
+    for key in sorted(params.keys()):
+        query_string += key + "=" + str(params[key]) + "&"
+    query_string = query_string[:-1]
+
+    return hmac.new(secret, query_string, hashlib.sha256).hexdigest()
+
+
+class BinanceJeExchange(ExchangeAPIWrapper):
+    currencies = None
+    credentials = None
+    base_url = "https://api.binance.je"
+    endpoints = {
+        "ping": {"req_method": "get", "url": "/api/v1/ping"},
+        "balance": {"req_method": "get", "url": "/api/v3/account"},
+        "prices": {"req_method": "get", "url": "/api/v3/ticker/price"},
+        "open_orders": {"req_method": "get", "url": "/api/v3/openOrders"},
+    }
+
+    def __init__(self, session=None, configuration=None):
+        if self.currencies is None:
+            raise NotImplementedError(
+                "Cannot instantiate BinanceJeExchange. Use a subclass instead."
+            )
+
+        super(BinanceJeExchange, self).__init__(session)
+        self.volume_currency = self.currencies["volume"]
+        self.currency = self.currencies["base"]
+        name_suffix = self.currencies["volume"] + "-" + self.currencies["base"]
+        self.name = "BINANCEJE_" + name_suffix
+        self.friendly_name = "Binance Jersey " + name_suffix
+
+    def load_credentials(self):
+        credentials = ["api_key", "secret"]
+        if self.credentials is None:
+            self.credentials = {
+                credential: self._load_env("BINANCEJE_" + credential.upper())
+                for credential in credentials
+            }
+
+    def auth_request(self, req_method, url, request_args):
+        """Modify a request to add authentication header and query string.
+
+        Overrides the not implemented method from the base class.
+
+        For authenticated endpoints, Binance requires the api key in the
+        request header and a timestamp and signature within the query string.
+        """
+        self.load_credentials()
+
+        try:
+            headers = request_args["headers"]
+        except KeyError:
+            headers = request_args["headers"] = {}
+
+        headers["X-MBX-APIKEY"] = self.credentials["api_key"]
+
+        try:
+            params = request_args["params"]
+        except KeyError:
+            params = request_args["params"] = {}
+
+        timestamp = int(round(time.time() * 1000))
+        params["timestamp"] = timestamp
+
+        signature = binance_signature(params, self.credentials["secret"])
+        params["signature"] = signature
+
+        logger.debug("headers: %s" % headers)
```

Comment:
Fine for now, when we're read to merge this we'll want to remove the log messages obviously.